// @wordpress/e2e-test-utils-playwright/build/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/e2e-test-utils-playwright/src/index.ts
var index_exports = {};
__export(index_exports, {
  Admin: () => import_admin.Admin,
  Editor: () => import_editor.Editor,
  Lighthouse: () => import_lighthouse.Lighthouse,
  Metrics: () => import_metrics.Metrics,
  PageUtils: () => import_page_utils.PageUtils,
  RequestUtils: () => import_request_utils.RequestUtils,
  expect: () => import_test.expect,
  test: () => import_test.test
});
module.exports = __toCommonJS(index_exports);
__reExport(index_exports, require("./types"), module.exports);
var import_admin = require("./admin");
var import_editor = require("./editor");
var import_page_utils = require("./page-utils");
var import_request_utils = require("./request-utils");
var import_metrics = require("./metrics");
var import_lighthouse = require("./lighthouse");
var import_test = require("./test");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Admin,
  Editor,
  Lighthouse,
  Metrics,
  PageUtils,
  RequestUtils,
  expect,
  test,
  ...require("./types")
});
