// @wordpress/interactivity/build/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "getConfig", {
  enumerable: true,
  get: function () {
    return _store.getConfig;
  }
});
Object.defineProperty(exports, "getContext", {
  enumerable: true,
  get: function () {
    return _scopes.getContext;
  }
});
Object.defineProperty(exports, "getElement", {
  enumerable: true,
  get: function () {
    return _scopes.getElement;
  }
});
Object.defineProperty(exports, "getServerContext", {
  enumerable: true,
  get: function () {
    return _scopes.getServerContext;
  }
});
Object.defineProperty(exports, "getServerState", {
  enumerable: true,
  get: function () {
    return _store.getServerState;
  }
});
exports.privateApis = void 0;
Object.defineProperty(exports, "splitTask", {
  enumerable: true,
  get: function () {
    return _utils.splitTask;
  }
});
Object.defineProperty(exports, "store", {
  enumerable: true,
  get: function () {
    return _store.store;
  }
});
Object.defineProperty(exports, "useCallback", {
  enumerable: true,
  get: function () {
    return _utils.useCallback;
  }
});
Object.defineProperty(exports, "useEffect", {
  enumerable: true,
  get: function () {
    return _utils.useEffect;
  }
});
Object.defineProperty(exports, "useInit", {
  enumerable: true,
  get: function () {
    return _utils.useInit;
  }
});
Object.defineProperty(exports, "useLayoutEffect", {
  enumerable: true,
  get: function () {
    return _utils.useLayoutEffect;
  }
});
Object.defineProperty(exports, "useMemo", {
  enumerable: true,
  get: function () {
    return _utils.useMemo;
  }
});
Object.defineProperty(exports, "useRef", {
  enumerable: true,
  get: function () {
    return _hooks2.useRef;
  }
});
Object.defineProperty(exports, "useState", {
  enumerable: true,
  get: function () {
    return _hooks2.useState;
  }
});
Object.defineProperty(exports, "useWatch", {
  enumerable: true,
  get: function () {
    return _utils.useWatch;
  }
});
Object.defineProperty(exports, "withScope", {
  enumerable: true,
  get: function () {
    return _utils.withScope;
  }
});
Object.defineProperty(exports, "withSyncEvent", {
  enumerable: true,
  get: function () {
    return _utils.withSyncEvent;
  }
});
var _preact = require("preact");
var _signals = require("@preact/signals");
var _directives = _interopRequireDefault(require("./directives"));
var _init = require("./init");
var _constants = require("./constants");
var _vdom = require("./vdom");
var _hooks = require("./hooks");
var _namespaces = require("./namespaces");
var _store = require("./store");
var _proxies = require("./proxies");
var _scopes = require("./scopes");
var _utils = require("./utils");
var _hooks2 = require("preact/hooks");
/**
 * External dependencies
 */

/**
 * Internal dependencies
 */

const requiredConsent = 'I acknowledge that using private APIs means my theme or plugin will inevitably break in the next version of WordPress.';
const privateApis = lock => {
  if (lock === requiredConsent) {
    return {
      directivePrefix: _constants.directivePrefix,
      getRegionRootFragment: _init.getRegionRootFragment,
      initialVdom: _init.initialVdom,
      toVdom: _vdom.toVdom,
      directive: _hooks.directive,
      getNamespace: _namespaces.getNamespace,
      h: _preact.h,
      cloneElement: _preact.cloneElement,
      render: _preact.render,
      proxifyState: _proxies.proxifyState,
      parseServerData: _store.parseServerData,
      populateServerData: _store.populateServerData,
      batch: _signals.batch
    };
  }
  throw new Error('Forbidden access.');
};
exports.privateApis = privateApis;
(0, _directives.default)();
(0, _init.init)();
